modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -68,4 +79,121 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
             transport.onException(tcpChannel, (Exception) newCause);
         }
     }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        assert msg instanceof ByteBuf;
+        final boolean queued = queuedWrites.offer(new WriteOperation((ByteBuf) msg, promise));
+        assert queued;
+    }
+
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+        if (ctx.channel().isWritable()) {
+            doFlush(ctx);
+        }
+        ctx.fireChannelWritabilityChanged();
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        Channel channel = ctx.channel();
+        if (channel.isWritable() || channel.isActive() == false) {
+            doFlush(ctx);
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        doFlush(ctx);
+        super.channelInactive(ctx);
+    }
+
+    private void resumeFlush() {
+        final ChannelHandlerContext context = this.ctx;
+        if (context == null) {
+            return;
+        }
+        if (context.executor().inEventLoop()) {
+            doFlush(context);
+        } else {
+            context.executor().execute(() -> doFlush(context));
+        }
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) {
+        this.ctx = ctx;
+    }
+
+    private void doFlush(ChannelHandlerContext ctx) {
+        final Channel channel = ctx.channel();
+        if (channel.isActive() == false) {
+            if (currentWrite != null) {
+                currentWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            WriteOperation queuedWrite;
+            while ((queuedWrite = queuedWrites.poll()) != null) {
+                queuedWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            return;
+        }
+        while (channel.isWritable()) {
+            if (currentWrite == null) {
+                currentWrite = queuedWrites.poll();
+            }
+            if (currentWrite == null) {
+                break;
+            }
+            final WriteOperation write = currentWrite;
```

Comment:
I think you will need to guard against the promise from already failing. The prior channel write could have failed and failed the the promise associated with the write operation. I imagine that this will kill the channel and cause everything to shutdown. But it still seems safest to prevent potentially flushing bytes that have already been released.

Reply:
That's why I put the cleanup loop that you mentioned [above](https://github.com/elastic/elasticsearch/pull/39286#discussion_r261421868). This way we def. fail all the writes and drain the queue on failures before going for another write in the loop. And since the only way, the callbacks/futures that would release the bytes is the `ctx.flush` we do at the end of the loop body I think we're safe here?